snuba/query/snql/parser.py

```diff
@@ -108,38 +141,78 @@ class OrTuple(NamedTuple):
     exp: Expression
 
 
+class EntityTuple(NamedTuple):
+    alias: str
+    name: str
+
+
 class SnQLVisitor(NodeVisitor):
     """
     Builds Snuba AST expressions from the Parsimonious parse tree.
     """
 
     def visit_query_exp(self, node: Node, visited_children: Iterable[Any]) -> Query:
-        _, where, collect, groupby, having, orderby = visited_children
+        match, where, collect, groupby, having, orderby, limit = visited_children
         # check for empty clauses
         if isinstance(groupby, Node):
             groupby = None
         if isinstance(orderby, Node):
             orderby = None
         if isinstance(having, Node):
             having = None
+        if isinstance(limit, Node):
+            limit = None
+
+        selected_columns = collect
+        _groupby = None
+        if groupby:
+            selected_columns += groupby
+            _groupby = [g.expression for g in groupby]
+
         return Query(
             body={},
-            from_clause=None,
-            selected_columns=collect,
+            from_clause=match,
+            selected_columns=selected_columns,
             array_join=None,
             condition=where,
             prewhere=None,
-            groupby=groupby,
+            groupby=_groupby,
             having=having,
             order_by=orderby,
+            limit=limit,
         )
 
+    def visit_match_clause(
+        self, node: Node, visited_children: Tuple[Any, Any, Any, EntityTuple, Any],
+    ) -> QueryEntity:
+        _, _, _, match, _ = visited_children
+        key = EntityKey(match.name)
+        query_entity = QueryEntity(key, get_entity(key).get_data_model())
+        return query_entity
+
+    def visit_entity_match(
+        self, node: Node, visited_children: Tuple[Any, str, Any, Any, str, Any],
+    ) -> EntityTuple:
+        _, alias, _, _, name, _ = visited_children
+        return EntityTuple(alias, name.lower())
```

Comment:
Any reason for not making entity name case sensitive in the lnaguage?  I don't think this constraint is applied anywhere when we define entities (it seems we kind of keep them lower case for convention). I would keep the language case sensitive for identifiers for two reasons:
- this would be the only case where we are permissive (all snuba and clickhouse function names and column names are going to be case sensitive).
- we would never be able to go back from this decision later because people may start querying `eVeNts` and that query will work. Basically the language would impose constraint on the evolution of the data model, which is undesirable, it should go the other way.

Reply:
In the language spec for SnQL we have the convention being `MATCH (e: Events)`, so I had to lower it to match the EntityKey `events`. If we want to make it case sensitive I can change the language specs.

Reply:
The specs were an example to me the actual names of data model identifiers does not have to match there since they are not part of the language but of the dataset. At that time we did not even have entities. I would change it to case sensitive, then up to you if you want to make the change in the spec.